rdtools/availability.py

```diff
@@ -0,0 +1,565 @@
+        resampled_cumulative = cumulative_energy.resample(rollup_period)
+        actual_production = (
+            resampled_cumulative.last() - resampled_cumulative.first()
+        )
+
+        lost_production = rdtools.energy_from_power(self.loss_total)
+        df = pd.DataFrame({
+            'lost_production': lost_production.resample(rollup_period).sum(),
+            'actual_production': actual_production,
+        })
+        loss_plus_actual = df['lost_production'] + df['actual_production']
+        df['availability'] = 1 - df['lost_production'] / loss_plus_actual
+        self.results = df.tz_localize(None)
+
+    def run(self, low_threshold=None, relative_sizes=None,
+            power_system_limit=None, quantiles=(0.01, 0.99),
+            rollup_period='m'):
+        """
+        Run the availability analysis.
+
+        Parameters
+        ----------
+        low_threshold : float or pd.Series, optional
+            An optional threshold used to naively classify subsystems as
+            online. If the threshold is a scalar, it will be used for all
+            subsystems. For subsystems with different capacities, a pandas
+            Series may be passed with index values matching the columns in
+            ``power_subsystem``. Units must match ``power_subsystem`` and
+            ``power_system``. If omitted, the limit is calculated for each
+            subsystem independently as 0.001 times the 99th percentile of its
+            power data.
+
+        relative_sizes : dict or pd.Series, optional
+            The production capacity of each subsystem, normalized by the mean
+            subsystem capacity. If not specified, it will be estimated from
+            power data.
+
+        power_system_limit : float or pd.Series, optional
+            An optional maximum system power used as an upper limit for
+            (power_system + lost_power) so that the maximum system capacity or
+            interconnection limit is not exceeded. If omitted, that check is
+            skipped.
+
+        quantiles : 2-element tuple, default (0.01, 0.99)
+            The quantiles of the error distribution used for the expected
+            energy confidence interval. The lower bound is used to classify
+            outages as either (1) a simple communication interruption with
+            no production loss or (2) a power outage with an associated
+            production loss estimate.
+
+        rollup_period : pandas offset string, default 'm'
+            The period on which to roll up losses and calculate availability.
+        """
+        self._loss_from_power(low_threshold, relative_sizes,
+                              power_system_limit)
+        self._error_distributions(quantiles)
+        self._loss_from_energy()
+        self._combine_losses(rollup_period)
+
+    def plot(self):
+        """
+        Create a figure summarizing the availability analysis results.
```

Comment:
Though it may be obvious to experienced python users, for someone without much experience in OOP, it may not be obvious that you need to run `AvailabilityAnalysis.run()` first. I suggest adding a note here and a descriptive exception in case someone tries to run this without having run the analysis first.